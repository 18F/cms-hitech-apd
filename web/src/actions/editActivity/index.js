export * from './contractorResources';
export * from './costAllocate';
export * from './overview';
