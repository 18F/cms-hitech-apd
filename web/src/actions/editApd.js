import { updateBudget } from './apd';

export const ADD_APD_ITEM = Symbol('add apd item');
export const EDIT_APD = Symbol('edit apd');
export const REMOVE_APD_ITEM = Symbol('remove apd item');

export const addKeyPerson = () => (dispatch, getState) => {
  dispatch({
    type: ADD_APD_ITEM,
    path: '/keyPersonnel/-',
    state: getState()
  });
};

export const removeKeyPerson = (
  index,
  { global = window } = {}
) => dispatch => {
  if (global.confirm('Do you really want to delete this key person?')) {
    dispatch({
      type: REMOVE_APD_ITEM,
      path: `/keyPersonnel/${index}`
    });
    dispatch(updateBudget());
  }
};

export const setKeyPersonName = (name, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/name`,
  value: name
});

export const setKeyPersonEmail = (email, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/email`,
  value: email
});

export const setKeyPersonRole = (role, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/position`,
  value: role
});

export const setKeyPersonPercentTime = (percent, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/percentTime`,
  value: percent
});

// needs to trigger budget calculation
export const setKeyPersonHasCosts = (hasCosts, index) => dispatch => {
  dispatch({
    type: EDIT_APD,
    path: `/keyPersonnel/${index}/hasCosts`,
    value: hasCosts
  });
  dispatch(updateBudget());
};

// needs to trigger budget calculation
export const setKeyPersonCost = (cost, index, year) => dispatch => {
  dispatch({
    type: EDIT_APD,
    path: `/keyPersonnel/${index}/costs/${year}`,
    value: cost
  });
  dispatch(updateBudget());
};

export const setMedicaidDirectorEmail = email => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidDirector/email',
    value: email
  };
};

export const setMedicaidDirectorName = name => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidDirector/name',
    value: name
  };
};

export const setMedicaidDirectorPhoneNumber = phone => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidDirector/phone',
    value: phone
  };
};

export const setMedicaidOfficeAddress1 = address => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidOffice/address1',
    value: address
  };
};

export const setMedicaidOfficeAddress2 = address => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidOffice/address2',
    value: address
  };
};

export const setMedicaidOfficeCity = city => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidOffice/city',
    value: city
  };
};

export const setMedicaidOfficeState = state => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidOffice/state',
    value: state
  };
};

export const setMedicaidOfficeZip = zip => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidOffice/zip',
    value: zip
  };
};
