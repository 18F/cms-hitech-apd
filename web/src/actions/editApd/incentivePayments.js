import { EDIT_APD } from './symbols';

export const setIncentiveEHCount = (year, quarter, value) => ({
  type: EDIT_APD,
  path: `/incentivePayments/ehCt/${year}/${quarter}`,
  value
});

export const setIncentiveEHPayment = (year, quarter, value) => ({
  type: EDIT_APD,
  path: `/incentivePayments/ehAmt/${year}/${quarter}`,
  value
});

export const setIncentiveEPCount = (year, quarter, value) => ({
  type: EDIT_APD,
  path: `/incentivePayments/epCt/${year}/${quarter}`,
  value
});

export const setIncentiveEPPayment = (year, quarter, value) => ({
  type: EDIT_APD,
  path: `/incentivePayments/epAmt/${year}/${quarter}`,
  value
});
