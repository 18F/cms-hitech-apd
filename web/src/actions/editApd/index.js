export * from './symbols';

export * from './keyPersonnel';
export * from './overview';
export * from './previousActivities';
export * from './stateProfile';
