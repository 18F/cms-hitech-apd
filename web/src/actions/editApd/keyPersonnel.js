import { ADD_APD_ITEM, EDIT_APD, REMOVE_APD_ITEM } from './symbols';
import { updateBudget } from '../apd';

export const addKeyPerson = () => (dispatch, getState) => {
  dispatch({
    type: ADD_APD_ITEM,
    path: '/keyPersonnel/-',
    state: getState()
  });
};

export const removeKeyPerson = (
  index,
  { global = window } = {}
) => dispatch => {
  if (global.confirm('Do you really want to delete this key person?')) {
    dispatch({
      type: REMOVE_APD_ITEM,
      path: `/keyPersonnel/${index}`
    });
    dispatch(updateBudget());
  }
};

export const setKeyPersonName = (name, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/name`,
  value: name
});

export const setKeyPersonEmail = (email, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/email`,
  value: email
});

export const setKeyPersonRole = (role, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/position`,
  value: role
});

export const setKeyPersonPercentTime = (percent, index) => ({
  type: EDIT_APD,
  path: `/keyPersonnel/${index}/percentTime`,
  value: percent
});

// needs to trigger budget calculation
export const setKeyPersonHasCosts = (hasCosts, index) => dispatch => {
  dispatch({
    type: EDIT_APD,
    path: `/keyPersonnel/${index}/hasCosts`,
    value: hasCosts
  });
  dispatch(updateBudget());
};

// needs to trigger budget calculation
export const setKeyPersonCost = (cost, index, year) => dispatch => {
  dispatch({
    type: EDIT_APD,
    path: `/keyPersonnel/${index}/costs/${year}`,
    value: cost
  });
  dispatch(updateBudget());
};
