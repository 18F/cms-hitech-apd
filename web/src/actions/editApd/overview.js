import { ADD_APD_YEAR, EDIT_APD, REMOVE_APD_YEAR } from './symbols';
import { updateBudget } from '../apd';

export const addYear = year => (dispatch, getState) => {
  dispatch({ type: ADD_APD_YEAR, value: year, state: getState() });
  dispatch(updateBudget());
};

export const removeYear = year => (dispatch, getState) => {
  dispatch({ type: REMOVE_APD_YEAR, value: year, state: getState() });
  dispatch(updateBudget());
};

export const setNarrativeForHIE = text => ({
  type: EDIT_APD,
  path: '/narrativeHIE',
  value: text
});

export const setNarrativeForHIT = text => ({
  type: EDIT_APD,
  path: '/narrativeHIT',
  value: text
});

export const setNarrativeForMMIS = text => ({
  type: EDIT_APD,
  path: '/narrativeMMIS',
  value: text
});

export const setProgramOverview = text => ({
  type: EDIT_APD,
  path: '/programOverview',
  value: text
});
