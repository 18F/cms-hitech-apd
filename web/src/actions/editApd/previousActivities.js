import { EDIT_APD } from './symbols';

export const setPreviousActivitySummary = summary => ({
  type: EDIT_APD,
  path: '/previousActivitySummary',
  value: summary
});

export const setPreviousActivityApprovedExpenseForHITandHIE = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/hithie/totalApproved`,
  value: expense
});

export const setPreviousActivityFederalActualExpenseForHITandHIE = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/hithie/federalActual`,
  value: expense
});

export const setPreviousActivityApprovedExpenseforMMIS50FFP = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/mmis/50/totalApproved`,
  value: expense
});

export const setPreviousActivityApprovedExpenseforMMIS75FFP = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/mmis/75/totalApproved`,
  value: expense
});

export const setPreviousActivityApprovedExpenseforMMIS90FFP = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/mmis/90/totalApproved`,
  value: expense
});

export const setPreviousActivityFederalActualExpenseforMMIS50FFP = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/mmis/50/federalActual`,
  value: expense
});

export const setPreviousActivityFederalActualExpenseforMMIS75FFP = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/mmis/75/federalActual`,
  value: expense
});

export const setPreviousActivityFederalActualExpenseforMMIS90FFP = (
  year,
  expense
) => ({
  type: EDIT_APD,
  path: `/previousActivityExpenses/${year}/mmis/90/federalActual`,
  value: expense
});
