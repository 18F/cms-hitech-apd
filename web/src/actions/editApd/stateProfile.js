import { EDIT_APD } from './symbols';

export const setMedicaidDirectorEmail = email => ({
  type: EDIT_APD,
  path: '/stateProfile/medicaidDirector/email',
  value: email
});

export const setMedicaidDirectorName = name => ({
  type: EDIT_APD,
  path: '/stateProfile/medicaidDirector/name',
  value: name
});

export const setMedicaidDirectorPhoneNumber = phone => ({
  type: EDIT_APD,
  path: '/stateProfile/medicaidDirector/phone',
  value: phone
});

export const setMedicaidOfficeAddress1 = address => ({
  type: EDIT_APD,
  path: '/stateProfile/medicaidOffice/address1',
  value: address
});

export const setMedicaidOfficeAddress2 = address => ({
  type: EDIT_APD,
  path: '/stateProfile/medicaidOffice/address2',
  value: address
});

export const setMedicaidOfficeCity = city => ({
  type: EDIT_APD,
  path: '/stateProfile/medicaidOffice/city',
  value: city
});

export const setMedicaidOfficeState = state => ({
  type: EDIT_APD,
  path: '/stateProfile/medicaidOffice/state',
  value: state
});

export const setMedicaidOfficeZip = zip => {
  return {
    type: EDIT_APD,
    path: '/stateProfile/medicaidOffice/zip',
    value: zip
  };
};
