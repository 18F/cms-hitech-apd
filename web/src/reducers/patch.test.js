import { SAVE_APD_SUCCESS } from '../actions/apd';
import { EDIT_APD } from '../actions/editApd';

import reducer, { getHasChanges } from './patch';

describe('JSON patch reducer', () => {
  it('provides an initial state', () => {
    expect(reducer(undefined, {})).toEqual([]);
  });

  it('resets to the initial state when an APD is saved', () => {
    expect(reducer({}, { type: SAVE_APD_SUCCESS })).toEqual([]);
  });

  it('pushes a new edit to the list when an APD is edited', () => {
    expect(
      reducer([], { type: EDIT_APD, path: '/path/to/edit', value: 'new value' })
    ).toEqual([{ op: 'replace', path: '/path/to/edit', value: 'new value' }]);
  });

  it('updates an existing patch if the path was replaced previously', () => {
    expect(
      reducer(
        [
          { op: 'replace', path: '/path/to/edit', value: 'old value' },
          { op: 'replace', path: '/path/to/leave', value: 'leave' }
        ],
        { type: EDIT_APD, path: '/path/to/edit', value: 'new value' }
      )
    ).toEqual([
      { op: 'replace', path: '/path/to/edit', value: 'new value' },
      { op: 'replace', path: '/path/to/leave', value: 'leave' }
    ]);
  });

  it('adds a new patch if the path was replaced previously but has had a different op applied since then', () => {
    expect(
      reducer(
        [
          { op: 'replace', path: '/path/to/edit', value: 'old value' },
          { op: 'remove', path: '/path/to/edit' }
        ],
        { type: EDIT_APD, path: '/path/to/edit', value: 'new value' }
      )
    ).toEqual([
      { op: 'replace', path: '/path/to/edit', value: 'old value' },
      { op: 'remove', path: '/path/to/edit' },
      { op: 'replace', path: '/path/to/edit', value: 'new value' }
    ]);
  });
});

describe('JSON patch selector', () => {
  it('returns false if there are no pending patches', () => {
    expect(getHasChanges({ patch: [] })).toEqual(false);
  });

  it('returns true if there are pending patches', () => {
    expect(getHasChanges({ patch: [''] })).toEqual(true);
  });
});
