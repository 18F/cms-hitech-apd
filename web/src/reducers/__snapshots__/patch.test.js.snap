// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON patch reducer adds a new year to the APD 1`] = `
Array [
  Object {
    "op": "replace",
    "path": "/years",
    "value": Array [
      "2019",
    ],
  },
  Object {
    "op": "add",
    "path": "/incentivePayments/ehAmt/2019",
    "value": Object {
      "1": 0,
      "2": 0,
      "3": 0,
      "4": 0,
    },
  },
  Object {
    "op": "add",
    "path": "/incentivePayments/ehCt/2019",
    "value": Object {
      "1": 0,
      "2": 0,
      "3": 0,
      "4": 0,
    },
  },
  Object {
    "op": "add",
    "path": "/incentivePayments/epAmt/2019",
    "value": Object {
      "1": 0,
      "2": 0,
      "3": 0,
      "4": 0,
    },
  },
  Object {
    "op": "add",
    "path": "/incentivePayments/epCt/2019",
    "value": Object {
      "1": 0,
      "2": 0,
      "3": 0,
      "4": 0,
    },
  },
  Object {
    "op": "add",
    "path": "/keyPersonnel/0/costs/2019",
    "value": 0,
  },
  Object {
    "op": "replace",
    "path": "/byKey/activityKey/years",
    "value": Array [
      "2019",
    ],
  },
  Object {
    "op": "add",
    "path": "/byKey/activityKey/contractorResources/0/hourly/data/2019",
    "value": Object {
      "hours": "",
      "rate": "",
    },
  },
  Object {
    "op": "add",
    "path": "/byKey/activityKey/contractorResources/0/years/2019",
    "value": 0,
  },
  Object {
    "op": "add",
    "path": "/byKey/activityKey/expenses/0/years/2019",
    "value": 0,
  },
  Object {
    "op": "add",
    "path": "/byKey/activityKey/statePersonnel/0/years/2019",
    "value": Object {
      "amt": "",
      "perc": "",
    },
  },
  Object {
    "op": "add",
    "path": "/byKey/activityKey/costAllocation/2019",
    "value": Object {
      "ffp": Object {
        "federal": 90,
        "state": 10,
      },
      "other": 0,
    },
  },
  Object {
    "op": "add",
    "path": "/byKey/activityKey/quarterlyFFP/2019",
    "value": Object {
      "1": Object {
        "combined": 25,
        "contractors": 25,
        "state": 25,
      },
      "2": Object {
        "combined": 25,
        "contractors": 25,
        "state": 25,
      },
      "3": Object {
        "combined": 25,
        "contractors": 25,
        "state": 25,
      },
      "4": Object {
        "combined": 25,
        "contractors": 25,
        "state": 25,
      },
    },
  },
]
`;

exports[`JSON patch reducer removes a year from the APD 1`] = `
Array [
  Object {
    "op": "remove",
    "path": "/years/0",
  },
  Object {
    "op": "remove",
    "path": "/incentivePayments/ehAmt/2019",
  },
  Object {
    "op": "remove",
    "path": "/incentivePayments/ehCt/2019",
  },
  Object {
    "op": "remove",
    "path": "/incentivePayments/epAmt/2019",
  },
  Object {
    "op": "remove",
    "path": "/incentivePayments/epCt/2019",
  },
  Object {
    "op": "remove",
    "path": "/keyPersonnel/0/costs/2019",
  },
  Object {
    "op": "remove",
    "path": "/byKey/activityKey/years/0",
  },
  Object {
    "op": "remove",
    "path": "/byKey/activityKey/contractorResources/0/hourly/data/2019",
  },
  Object {
    "op": "remove",
    "path": "/byKey/activityKey/contractorResources/0/years/2019",
  },
  Object {
    "op": "remove",
    "path": "/byKey/activityKey/expenses/0/years/2019",
  },
  Object {
    "op": "remove",
    "path": "/byKey/activityKey/statePersonnel/0/years/2019",
  },
  Object {
    "op": "remove",
    "path": "/byKey/activityKey/costAllocation/2019",
  },
  Object {
    "op": "remove",
    "path": "/byKey/activityKey/quarterlyFFP/2019",
  },
]
`;
